### ISSUE TITLE
Flow does not complain about an undefined type (Dispatch)

### ISSUE THREAD

**Issue by NameFILIP** [facebook/flow#6876]
The type is defined in the [libdef for redux](https://github.com/flow-typed/flow-typed/blob/master/definitions/npm/redux_v4.x.x/flow_v0.55.x-/redux_v4.x.x.js#L12):
```
declare export type DispatchAPI<A> = (action: A) => A;
declare export type Dispatch<A: { type: $Subtype<string> }> = DispatchAPI<A>;
```

In the component I define a function:
```
type Action = { type: 'Action1' } | { type: 'Action2' };

function mapDispatchToProps(dispatch: Dispatch<Action>) {
  return {};
}
```

If I import the type everything works as expected:
```
import type { Dispatch } from 'redux';
```

But if I forget to import it, `Dispatch` becomes `any` and allows any action type which leads to bugs in the application if a shape of an action is incorrect.

* If I comment out the `Dispatch` definition in the libdef, the type of `Dispatch` agrument changes from `any` to `empty`.
* If I rename `Dispatch` to any other variable, like `ReduxDispatch`, flowtype starts complaining.
* If I rename `Dispatch` to another variable defined in libdef, like `MiddlewareAPI` it becomes `any` and does not complain.

Node v8.11.4
Flow v0.80.0

**Question:** Why does flowtype make `Dispatch` = `any` instead of complaining about a missing definition?

**Comment by wchargin** [comment]
Please describe how to reproduce the problem.

I created a file `test.js` with the following contents…

```js
// @flow
type Action = { type: "Action1" } | { type: "Action2" };
function mapDispatchToProps(dispatch: Dispatch<Action>) {
  return {};
}
```

…then installed Node v8.11.4 and Flow v0.80.0, and Flow properly
reported the unresolved type:

```
$ node --version
v8.11.4
$ node ./node_modules/.bin/flow version
Flow, a static type checker for JavaScript, version 0.80.0
$ shasum -a 256 ./test.js
f87d8bae843979887d730629a270c4723e9356e332f55eda9eafbb017ed1c22a  ./test.js
$ node ./node_modules/.bin/flow
Error ┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ test.js:3:39

Cannot resolve name Dispatch.

     1│ // @flow
     2│ type Action = { type: "Action1" } | { type: "Action2" };
     3│ function mapDispatchToProps(dispatch: Dispatch<Action>) {
     4│   return {};
     5│ }
     6│



Found 1 error
$ uname -sr; lsb_release -d
Linux 4.10.0-32-generic
Description:    Linux Mint 18.2 Sonya
```

The same behavior occurs if I `flow-typed install redux@4.x.x` and
re-run Flow.

The above `test.js` file also fails properly on <https://flow.org/try/>.

**Comment by NameFILIP** [comment]
I traced it down to the `react-redux_v5.x.x` flow-typed definition. Looks like it's already fixed on the most recent version, ours was a bit old. Here is the content of ours:

```
// flow-typed signature: 59b0c4be0e1408f21e2446be96c79804
// flow-typed version: 9092387fd2/react-redux_v5.x.x/flow_>=v0.54.x

import type { Dispatch, Store } from "redux";

declare module "react-redux" {
  /*

    S = State
    A = Action
    OP = OwnProps
    SP = StateProps
    DP = DispatchProps

  */

  declare type MapStateToProps<S, OP: Object, SP: Object> = (
    state: S,
    ownProps: OP
  ) => ((state: S, ownProps: OP) => SP) | SP;

  declare type MapDispatchToProps<A, OP: Object, DP: Object> =
    | ((dispatch: Dispatch<A>, ownProps: OP) => DP)
    | DP;

  declare type MergeProps<SP, DP: Object, OP: Object, P: Object> = (
    stateProps: SP,
    dispatchProps: DP,
    ownProps: OP
  ) => P;

  declare type Context = { store: Store<*, *> };

  declare type ComponentWithDefaultProps<DP: {}, P: {}, CP: P> = Class<
    React$Component<CP>
  > & { defaultProps: DP };

  declare class ConnectedComponentWithDefaultProps<
    OP,
    DP,
    CP
  > extends React$Component<OP> {
    static defaultProps: DP, // <= workaround for https://github.com/facebook/flow/issues/4644
    static WrappedComponent: Class<React$Component<CP>>,
    getWrappedInstance(): React$Component<CP>,
    props: OP,
    state: void
  }

  declare class ConnectedComponent<OP, P> extends React$Component<OP> {
    static WrappedComponent: Class<React$Component<P>>,
    getWrappedInstance(): React$Component<P>,
    props: OP,
    state: void
  }

  declare type ConnectedComponentWithDefaultPropsClass<OP, DP, CP> = Class<
    ConnectedComponentWithDefaultProps<OP, DP, CP>
  >;

  declare type ConnectedComponentClass<OP, P> = Class<
    ConnectedComponent<OP, P>
  >;

  declare type Connector<OP, P> = (<DP: {}, CP: {}>(
    component: ComponentWithDefaultProps<DP, P, CP>
  ) => ConnectedComponentWithDefaultPropsClass<OP, DP, CP>) &
    ((component: React$ComponentType<P>) => ConnectedComponentClass<OP, P>);

  declare class Provider<S, A> extends React$Component<{
    store: Store<S, A>,
    children?: any
  }> {}

  declare function createProvider(
    storeKey?: string,
    subKey?: string
  ): Provider<*, *>;

  declare type ConnectOptions = {
    pure?: boolean,
    withRef?: boolean
  };

  declare type Null = null | void;

  declare function connect<A, OP>(
    ...rest: Array<void> // <= workaround for https://github.com/facebook/flow/issues/2360
  ): Connector<OP, $Supertype<{ dispatch: Dispatch<A> } & OP>>;

  declare function connect<A, OP>(
    mapStateToProps: Null,
    mapDispatchToProps: Null,
    mergeProps: Null,
    options: ConnectOptions
  ): Connector<OP, $Supertype<{ dispatch: Dispatch<A> } & OP>>;

  declare function connect<S, A, OP, SP>(
    mapStateToProps: MapStateToProps<S, OP, SP>,
    mapDispatchToProps: Null,
    mergeProps: Null,
    options?: ConnectOptions
  ): Connector<OP, $Supertype<SP & { dispatch: Dispatch<A> } & OP>>;

  declare function connect<A, OP, DP>(
    mapStateToProps: Null,
    mapDispatchToProps: MapDispatchToProps<A, OP, DP>,
    mergeProps: Null,
    options?: ConnectOptions
  ): Connector<OP, $Supertype<DP & OP>>;

  declare function connect<S, A, OP, SP, DP>(
    mapStateToProps: MapStateToProps<S, OP, SP>,
    mapDispatchToProps: MapDispatchToProps<A, OP, DP>,
    mergeProps: Null,
    options?: ConnectOptions
  ): Connector<OP, $Supertype<SP & DP & OP>>;

  declare function connect<S, A, OP, SP, DP, P>(
    mapStateToProps: MapStateToProps<S, OP, SP>,
    mapDispatchToProps: Null,
    mergeProps: MergeProps<SP, DP, OP, P>,
    options?: ConnectOptions
  ): Connector<OP, P>;

  declare function connect<S, A, OP, SP, DP, P>(
    mapStateToProps: MapStateToProps<S, OP, SP>,
    mapDispatchToProps: MapDispatchToProps<A, OP, DP>,
    mergeProps: MergeProps<SP, DP, OP, P>,
    options?: ConnectOptions
  ): Connector<OP, P>;
}
```

**Comment by NameFILIP** [comment]
yeah, see the comment: https://github.com/flow-typed/flow-typed/blob/master/definitions/npm/react-redux_v5.x.x/flow_v0.68.0-/react-redux_v5.x.x.js#L4

**Comment by wchargin** [comment]
Ah. This is #6650, then; see also <https://github.com/flow-typed/flow-typed/blob/master/CONTRIBUTING.md#dont-import-types-from-other-libdefs>.

**Comment by aij** [comment]
> Flow silently converts `MyType` to be typed `any`, and then sadness ensues.

The sadness that ensues seems to be a lot worse than just importing it as type `any`, as you might expect from that comment.

**Comment by gnprice** [comment]
This was fixed in Flow v0.144 (perhaps earlier), so this issue can be closed.

The story is the same as at https://github.com/facebook/flow/issues/6414#issuecomment-958648596 . The react-redux libdef above had the following line at toplevel, outside of any `declare module` block:
```
import type { Dispatch, Store } from "redux";
```

Flow will now reject that with the following error:
> Cannot use an import statement at the toplevel of a library file. Import statements may only appear inside a `declare module`. [toplevel-library-import]

If you put the import statement inside a `declare module` block instead, all works great.
